### ISSUE TITLE
Auto-load configuration file in home directory if no project-specific configuration found

### ISSUE THREAD

**Issue by iainbeeston** [brigade/scss-lint#436]
Would it be possible to check for a .scss-lint.yml file in the user's home directory? So that it looks for config in the following order:

1. Working directory
2. Home directory
3. scss-lint config directory

For most linting tools (rubocop, jshint) I keep a config file in my home directory which is the default for all projects I work on, but it seems that scss-lint only looks in the working directory for config.

If you can give me a few directions I might be able to implement this myself.

**Comment by chris-dura** [comment]
Can't you just pass a config file by using the `--config` argument in the command...
```bash
scss-lint project/**/*.scss --config ~/your/home/directory/scss-lint.yml
```

**Comment by iainbeeston** [comment]
I use scss-lint via my editor, so it's harder to pass through options.

Also an explicit option is harder to remember. With other linting tools it
happens automatically, so I can set options once and then forget about them.

**Comment by sds** [comment]
Hey @iainbeeston,

I understand the desire here, but it is slightly at odds with how we believe this tool should be used.

Ideally, you should have a `.scss-lint.yml` file for each of the repositories you lint with `scss-lint`. This ensures that anyone else working on your repository also uses the same configuration, as it is kept up to date with the repo.

I understand that might be perceived as a pain if you have multiple repos, but the benefits are pretty significant once you realize that you get the distribution of the configuration for free when you do this: regardless of whether you are running in a CI environment or on another developer's machine, the configuration is the same and you don't need to ensure you're keeping all those locations in sync.

Does that make sense?

**Comment by iainbeeston** [comment]
Yes it makes sense and I'm familiar with that viewpoint. Unfortunately not
every developer on my project uses scss-lint. I see it as part of my own
personal setup, and beyond the most fundamental config (like indentation
level) I prefer to keep editor or linting config separate from the source
control I use for code.

There are a few reasons that I like to work this way. I don't want other
developers to be lumbered with config they don't use (or even understand).
Files that are checked into source control alongside the code have a habit
of ending up deployed on production servers. It also feels like a better
separation of concerns to me - editor, linting and build config are each
only useful on one environment.

Other linting tools allow both per-project and per-user config (as does
git, and other Unix tools). I could check .scss-lint.yml into my project
but at present it's the only tool I use that has this issue. The best
compromise I can think of is to include it in my .gitignore file but that
still means I'll need to copy or symlink the file into every project that I
use scss on.

If it's not a feature you guys want then so be it: I'll have to work around
it in future.

**Comment by sds** [comment]
I'm open to accepting a pull request adding support for this (I don't think it would be hard to implement) if you want this sooner rather than later; otherwise I'll get to it when I can. I understand this is relatively standard with other UNIX tools, and if developers would rather use a generic user-level configuration in their home directory then we should let them.

Some thoughts on your points, however:

> Unfortunately not every developer on my project uses scss-lint.

Then I think you're doing it wrong! :stuck_out_tongue: A huge advantage here is enforcing consistency amongst your team for how to write SCSS. I understand that getting your team to agree on a set of standards can be a hurdle, but once done it ends up saving you time in the long run as you can now enforce these conventions in an automated way, instead of via code review or ad hoc discussion each time it comes up.

> I prefer to keep editor or linting config separate from the source control I use for code.

Editor config I agree with here to some extent (no one stores their `.vimrc` in source control for projects), but projects like [EditorConfig](http://editorconfig.org/) aim to provide a standard way to enforce things like tabs vs spaces etc. (most editors have a plugin to enable support), which is also something that should be linked to the repository. Having different developers using different tab/space results in unnecessary code changes when editing the same file (introducing the possibility of merge conflicts etc.)

Similarly with linting: why would you want to vary your lint configuration from person to person? That introduces inconsistency that can cause unnecessary work. If one developer prefers selectors with format X and another format Y, they'll keep changing each other's code when working on the same set of files. Storing it with the repository is the easiest way of guaranteeing this enforcement&mdash;you don't need to ensure developers install another repo with all of your standardized lint configuration.

> I don't want other developers to be lumbered with config they don't use

I don't understand the burden you're implying here. It's a file that doesn't affect anyone unless they choose to interact with it by running `scss-lint`.

> Files that are checked into source control alongside the code have a habit of ending up deployed on production servers

Can you elaborate on why this is a bad thing? You're not shipping an executable, you're shipping a small text file. That's not going to break the bandwidth bank on any deploy process, and unless you're doing something really funky it shouldn't result in that file being served by your web server.

> editor, linting and build config are each only useful on one environment.

This depends on your development process. At Brigade, we use [Overcommit](https://github.com/brigade/overcommit) to run git hooks in development so that linters can run against code before it is even committed. We then double-check in CI test environment that these hooks actually ran by running the linters again, but this time against the entire repository. This is certainly a level up from what most development teams do, but we believe the consistency it provides (and the incredibly low cost to implement) makes it worthwhile.

At the end of the day, I understand different teams will have differing opinions with how they want to develop. Since Brigade maintains `scss-lint`, it is necessarily influenced by the philosophy suggested above, and so some features will lend themselves towards supporting that philosophy. Hopefully this doesn't preclude different uses, so your request is still a valid one.

**Comment by iainbeeston** [comment]
Thanks for the detailed response. I'll see if I can compose a PR later this
week and see what people think the changes involved.
